File: app/rust/src/ch04_higher_order.rs
//! 第4章: 関数を値として扱う
//!
//! 高階関数（Higher-Order Function）を学びます。
//! 関数を引数として受け取る、または関数を戻り値として返す関数です。

// =============================================================================
// 4.1 高階関数とは
// =============================================================================

/// 高階関数の基本: 関数を引数として受け取る
///
/// # Examples
///
/// ```
/// use grokking_fp::ch04_higher_order::apply_twice;
///
/// fn double(x: i32) -> i32 { x * 2 }
///
/// assert_eq!(apply_twice(double, 3), 12); // double(double(3)) = double(6) = 12
/// ```
pub fn apply_twice<T, F>(f: F, x: T) -> T
where
    F: Fn(T) -> T,
{
    f(f(x))
}

// =============================================================================
// 4.2 関数を引数として渡す
// =============================================================================

/// 'a' を除いた文字数でスコアを計算
///
/// # Examples
///
/// ```
/// use grokking_fp::ch04_higher_order::score;
///
/// assert_eq!(score("rust"), 4);  // r, u, s, t
/// assert_eq!(score("java"), 2);  // j, v (a を除く)
/// assert_eq!(score("scala"), 3); // s, c, l (a を除く)
/// ```
pub fn score(word: &str) -> usize {
    word.chars().filter(|&c| c != 'a').count()
}

/// スコア順にソート（昇順）
///
/// # Examples
///
/// ```
/// use grokking_fp::ch04_higher_order::{score, sort_by_score};
///
/// let words = vec!["rust", "java"];
/// let sorted = sort_by_score(&words, |w| score(w));
///
/// assert_eq!(sorted, vec!["java", "rust"]);
/// // java: 2, rust: 4
/// ```
pub fn sort_by_score<T, F, K>(list: &[T], key_fn: F) -> Vec<T>
where
    T: Clone,
    F: Fn(&T) -> K,
    K: Ord,
{
    let mut result = list.to_vec();
    result.sort_by_key(|item| key_fn(item));
    result
}

/// map: 各要素を変換
///
/// # Examples
///
/// ```
/// use grokking_fp::ch04_higher_order::map;
///
/// let lengths: Vec<usize> = map(&["scala", "rust", "ada"], |s| s.len());
/// assert_eq!(lengths, vec![5, 4, 3]);
///
/// let doubles: Vec<i32> = map(&[5, 1, 2, 4, 0], |&i| i * 2);
/// assert_eq!(doubles, vec![10, 2, 4, 8, 0]);
/// ```
pub fn map<T, U, F>(list: &[T], f: F) -> Vec<U>
where
    F: Fn(&T) -> U,
{
    list.iter().map(f).collect()
}

/// filter: 条件に合う要素を抽出
///
/// # Examples
///
/// ```
/// use grokking_fp::ch04_higher_order::filter;
///
/// let odds: Vec<i32> = filter(&[5, 1, 2, 4, 0], |&i| i % 2 == 1);
/// assert_eq!(odds, vec![5, 1]);
///
/// let large: Vec<i32> = filter(&[5, 1, 2, 4, 0], |&i| i > 4);
/// assert_eq!(large, vec![5]);
/// ```
pub fn filter<T: Clone, F>(list: &[T], predicate: F) -> Vec<T>
where
    F: Fn(&T) -> bool,
{
    list.iter().filter(|item| predicate(item)).cloned().collect()
}

/// fold: 畳み込み（foldLeft 相当）
///
/// # Examples
///
/// ```
/// use grokking_fp::ch04_higher_order::fold;
///
/// let sum = fold(&[5, 1, 2, 4, 100], 0, |acc, &i| acc + i);
/// assert_eq!(sum, 112);
///
/// let max = fold(&[5, 1, 2, 4, 15], i32::MIN, |max, &i| if i > max { i } else { max });
/// assert_eq!(max, 15);
///
/// let product = fold(&[1, 2, 3, 4], 1, |acc, &i| acc * i);
/// assert_eq!(product, 24);
/// ```
pub fn fold<T, U, F>(list: &[T], initial: U, f: F) -> U
where
    F: Fn(U, &T) -> U,
{
    list.iter().fold(initial, f)
}

// =============================================================================
// 4.3 構造体とパターン
// =============================================================================

/// プログラミング言語を表す構造体
#[derive(Debug, Clone, PartialEq)]
pub struct ProgrammingLanguage {
    pub name: String,
    pub year: i32,
}

impl ProgrammingLanguage {
    pub fn new(name: &str, year: i32) -> Self {
        Self {
            name: name.to_string(),
            year,
        }
    }
}

/// 構造体のフィールドを使った操作
///
/// # Examples
///
/// ```
/// use grokking_fp::ch04_higher_order::{ProgrammingLanguage, get_names, filter_by_year};
///
/// let java = ProgrammingLanguage::new("Java", 1995);
/// let scala = ProgrammingLanguage::new("Scala", 2004);
/// let languages = vec![java.clone(), scala.clone()];
///
/// let names = get_names(&languages);
/// assert_eq!(names, vec!["Java", "Scala"]);
///
/// let young = filter_by_year(&languages, 2000);
/// assert_eq!(young, vec![scala]);
/// ```
pub fn get_names(languages: &[ProgrammingLanguage]) -> Vec<String> {
    languages.iter().map(|lang| lang.name.clone()).collect()
}

pub fn filter_by_year(languages: &[ProgrammingLanguage], min_year: i32) -> Vec<ProgrammingLanguage> {
    languages
        .iter()
        .filter(|lang| lang.year > min_year)
        .cloned()
        .collect()
}

// =============================================================================
// 4.4 関数を返す関数
// =============================================================================

/// n より大きいかを判定する関数を返す
///
/// # Examples
///
/// ```
/// use grokking_fp::ch04_higher_order::larger_than;
///
/// let larger_than_4 = larger_than(4);
/// assert!(larger_than_4(5));
/// assert!(!larger_than_4(4));
/// assert!(!larger_than_4(3));
///
/// // filter で使う場合は参照を受け取るクロージャでラップ
/// let large: Vec<i32> = vec![5, 1, 2, 4, 0]
///     .into_iter()
///     .filter(|x| larger_than(4)(*x))
///     .collect();
/// assert_eq!(large, vec![5]);
/// ```
pub fn larger_than(n: i32) -> impl Fn(i32) -> bool {
    move |i| i > n
}

/// n より大きいかを判定する関数を返す（参照版）
///
/// # Examples
///
/// ```
/// use grokking_fp::ch04_higher_order::larger_than_ref;
///
/// let nums = vec![5, 1, 2, 4, 0];
/// let large: Vec<&i32> = nums.iter().filter(larger_than_ref(4)).collect();
/// assert_eq!(large, vec![&5]);
///
/// let medium: Vec<&i32> = nums.iter().filter(larger_than_ref(1)).collect();
/// assert_eq!(medium, vec![&5, &2, &4]);
/// ```
pub fn larger_than_ref(n: i32) -> impl Fn(&&i32) -> bool {
    move |&&i| i > n
}

/// n で割り切れるかを判定する関数を返す
///
/// # Examples
///
/// ```
/// use grokking_fp::ch04_higher_order::divisible_by;
///
/// let divisible_by_3 = divisible_by(3);
/// assert!(divisible_by_3(9));
/// assert!(divisible_by_3(6));
/// assert!(!divisible_by_3(7));
/// ```
pub fn divisible_by(n: i32) -> impl Fn(i32) -> bool {
    move |i| i % n == 0
}

/// 文字を含むかを判定する関数を返す
///
/// # Examples
///
/// ```
/// use grokking_fp::ch04_higher_order::contains_char;
///
/// let contains_s = contains_char('s');
/// assert!(contains_s("rust"));
/// assert!(contains_s("scala"));
/// assert!(!contains_s("java"));
/// ```
pub fn contains_char(c: char) -> impl Fn(&str) -> bool {
    move |s| s.contains(c)
}

// =============================================================================
// 4.5 カリー化
// =============================================================================

/// カリー化された関数: 2つの数を比較
///
/// # Examples
///
/// ```
/// use grokking_fp::ch04_higher_order::curried_larger;
///
/// let larger_than_10 = curried_larger(10);
/// assert!(larger_than_10(15));
/// assert!(!larger_than_10(5));
/// ```
pub fn curried_larger(threshold: i32) -> impl Fn(i32) -> bool {
    move |value| value > threshold
}

/// カリー化された加算
///
/// # Examples
///
/// ```
/// use grokking_fp::ch04_higher_order::curried_add;
///
/// let add_10 = curried_add(10);
/// assert_eq!(add_10(5), 15);
/// assert_eq!(add_10(20), 30);
/// ```
pub fn curried_add(a: i32) -> impl Fn(i32) -> i32 {
    move |b| a + b
}

// =============================================================================
// 4.6 ワードスコアリング
// =============================================================================

/// ボーナススコア: 'c' を含む場合は +5
///
/// # Examples
///
/// ```
/// use grokking_fp::ch04_higher_order::bonus;
///
/// assert_eq!(bonus("scala"), 5);
/// assert_eq!(bonus("clojure"), 5);
/// assert_eq!(bonus("rust"), 0);
/// ```
pub fn bonus(word: &str) -> i32 {
    if word.contains('c') { 5 } else { 0 }
}

/// ペナルティ: 's' を含む場合は -7
///
/// # Examples
///
/// ```
/// use grokking_fp::ch04_higher_order::penalty;
///
/// assert_eq!(penalty("scala"), 7);
/// assert_eq!(penalty("rust"), 7);
/// assert_eq!(penalty("java"), 0);
/// ```
pub fn penalty(word: &str) -> i32 {
    if word.contains('s') { 7 } else { 0 }
}

/// スコアリング関数を使ってランキングを作成
///
/// # Examples
///
/// ```
/// use grokking_fp::ch04_higher_order::{score, bonus, penalty, ranked_words};
///
/// let words = vec!["ada", "haskell", "scala", "java", "rust"];
///
/// // 基本スコアでランキング
/// let ranking1 = ranked_words(&words, |w| score(w) as i32);
/// assert_eq!(ranking1, vec!["haskell", "rust", "scala", "java", "ada"]);
///
/// // ボーナス付きスコアでランキング
/// let ranking2 = ranked_words(&words, |w| score(w) as i32 + bonus(w));
/// assert_eq!(ranking2, vec!["scala", "haskell", "rust", "java", "ada"]);
///
/// // ボーナスとペナルティ付きスコアでランキング
/// // java: 2+0-0=2, scala: 3+5-7=1, ada: 0+0-0=0, haskell: 6+0-7=-1, rust: 4+0-7=-3
/// let ranking3 = ranked_words(&words, |w| score(w) as i32 + bonus(w) - penalty(w));
/// assert_eq!(ranking3[0], "java");   // 2点
/// assert_eq!(ranking3[4], "rust");   // -3点
/// ```
pub fn ranked_words<'a, F>(words: &[&'a str], word_score: F) -> Vec<&'a str>
where
    F: Fn(&str) -> i32,
{
    let mut result: Vec<&'a str> = words.to_vec();
    result.sort_by(|a, b| word_score(b).cmp(&word_score(a))); // 降順
    result
}

/// 条件を満たす要素の数をカウント
///
/// # Examples
///
/// ```
/// use grokking_fp::ch04_higher_order::count_where;
///
/// assert_eq!(count_where(&[1, 2, 3, 4, 5], |&i| i > 3), 2);
/// assert_eq!(count_where(&["a", "bb", "ccc"], |s| s.len() > 1), 2);
/// ```
pub fn count_where<T, F>(list: &[T], predicate: F) -> usize
where
    F: Fn(&T) -> bool,
{
    list.iter().filter(|item| predicate(item)).count()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_apply_twice() {
        let double = |x: i32| x * 2;
        assert_eq!(apply_twice(double, 3), 12);
        assert_eq!(apply_twice(double, 5), 20);
    }

    #[test]
    fn test_score() {
        assert_eq!(score("rust"), 4);
        assert_eq!(score("java"), 2);
        assert_eq!(score("scala"), 3);
        assert_eq!(score("aaa"), 0);
    }

    #[test]
    fn test_sort_by_score() {
        let words = vec!["rust", "java", "ada"];
        let sorted = sort_by_score(&words, |w| score(w));
        assert_eq!(sorted, vec!["ada", "java", "rust"]);
    }

    #[test]
    fn test_map() {
        let nums = vec![1, 2, 3];
        let doubled: Vec<i32> = map(&nums, |&x| x * 2);
        assert_eq!(doubled, vec![2, 4, 6]);
    }

    #[test]
    fn test_filter() {
        let nums = vec![1, 2, 3, 4, 5];
        let evens: Vec<i32> = filter(&nums, |&x| x % 2 == 0);
        assert_eq!(evens, vec![2, 4]);
    }

    #[test]
    fn test_fold() {
        let nums = vec![1, 2, 3, 4, 5];
        let sum = fold(&nums, 0, |acc, &x| acc + x);
        assert_eq!(sum, 15);

        let concat = fold(&["a", "b", "c"], String::new(), |acc, &s| acc + s);
        assert_eq!(concat, "abc");
    }

    #[test]
    fn test_programming_language() {
        let java = ProgrammingLanguage::new("Java", 1995);
        let scala = ProgrammingLanguage::new("Scala", 2004);
        let rust = ProgrammingLanguage::new("Rust", 2010);
        let languages = vec![java, scala.clone(), rust.clone()];

        let names = get_names(&languages);
        assert_eq!(names, vec!["Java", "Scala", "Rust"]);

        let young = filter_by_year(&languages, 2000);
        assert_eq!(young, vec![scala, rust]);
    }

    #[test]
    fn test_larger_than() {
        let larger_than_5 = larger_than(5);
        assert!(larger_than_5(10));
        assert!(!larger_than_5(5));
        assert!(!larger_than_5(3));
    }

    #[test]
    fn test_divisible_by() {
        let div_by_2 = divisible_by(2);
        assert!(div_by_2(4));
        assert!(div_by_2(10));
        assert!(!div_by_2(5));
    }

    #[test]
    fn test_contains_char() {
        let has_r = contains_char('r');
        assert!(has_r("rust"));
        assert!(!has_r("java"));
    }

    #[test]
    fn test_curried_add() {
        let add_5 = curried_add(5);
        assert_eq!(add_5(10), 15);
        assert_eq!(add_5(0), 5);
    }

    #[test]
    fn test_bonus_penalty() {
        assert_eq!(bonus("scala"), 5);
        assert_eq!(bonus("rust"), 0);
        assert_eq!(penalty("scala"), 7);
        assert_eq!(penalty("java"), 0);
    }

    #[test]
    fn test_ranked_words() {
        let words = vec!["ada", "haskell", "scala", "java", "rust"];

        // 基本スコア
        let ranking1 = ranked_words(&words, |w| score(w) as i32);
        assert_eq!(ranking1[0], "haskell"); // 7文字、aなし

        // ボーナス付き
        let ranking2 = ranked_words(&words, |w| score(w) as i32 + bonus(w));
        assert_eq!(ranking2[0], "scala"); // 3 + 5 = 8
    }

    #[test]
    fn test_count_where() {
        assert_eq!(count_where(&[1, 2, 3, 4, 5], |&x| x > 3), 2);
        assert_eq!(count_where(&[1, 2, 3, 4, 5], |&x| x % 2 == 0), 2);
    }
}

File: app/typescript/src/ch04_higher_order.ts
/**
 * 第4章: 関数を値として扱う
 *
 * 関数を他の関数に渡したり、関数から関数を返したりする方法を学びます。
 * TypeScript では関数は第一級オブジェクトなので、変数に代入したり、
 * 引数として渡したりすることができます。
 */

import { pipe, flow } from 'fp-ts/function'
import * as RA from 'fp-ts/ReadonlyArray'
import * as Ord from 'fp-ts/Ord'
import * as N from 'fp-ts/number'
import * as S from 'fp-ts/string'

// =============================================================================
// 4.1 関数を引数として渡す（sortBy、map、filter）
// =============================================================================

/**
 * 単語のスコアを計算する（'a'を除いた文字数）。
 *
 * @example
 * score("java") // => 2
 * score("rust") // => 4
 */
export const score = (word: string): number =>
  word.replace(/a/gi, '').length

/**
 * 単語をスコア順（降順）にソートする。
 *
 * @example
 * rankedWords(["rust", "java"], score) // => ["rust", "java"]
 * rankedWords(["ada", "haskell", "scala"], score) // => ["haskell", "scala", "ada"]
 */
export const rankedWords = (
  words: readonly string[],
  wordScore: (word: string) => number
): readonly string[] =>
  pipe(
    words,
    RA.sortBy([Ord.reverse(Ord.contramap(wordScore)(N.Ord))])
  )

/**
 * 文字列を長さ順（昇順）にソートする。
 *
 * @example
 * sortByLength(["scala", "rust", "ada"]) // => ["ada", "rust", "scala"]
 */
export const sortByLength = (words: readonly string[]): readonly string[] =>
  pipe(
    words,
    RA.sortBy([Ord.contramap((s: string) => s.length)(N.Ord)])
  )

/**
 * 文字列を長さ順（降順）にソートする。
 *
 * @example
 * sortByLengthDesc(["scala", "rust", "ada"]) // => ["scala", "rust", "ada"]
 */
export const sortByLengthDesc = (words: readonly string[]): readonly string[] =>
  pipe(
    words,
    RA.sortBy([Ord.reverse(Ord.contramap((s: string) => s.length)(N.Ord))])
  )

/**
 * 文字列中の's'の数を数える。
 *
 * @example
 * numberOfS("rust") // => 1
 * numberOfS("haskell") // => 1
 * numberOfS("ada") // => 0
 */
export const numberOfS = (s: string): number =>
  s.split('').filter((c) => c === 's').length

/**
 * 文字列を's'の数順（昇順）にソートする。
 *
 * @example
 * sortByNumberOfS(["rust", "ada"]) // => ["ada", "rust"]
 */
export const sortByNumberOfS = (words: readonly string[]): readonly string[] =>
  pipe(
    words,
    RA.sortBy([Ord.contramap(numberOfS)(N.Ord)])
  )

// =============================================================================
// 4.2 map - 各要素に関数を適用
// =============================================================================

/**
 * 各文字列の長さを取得する。
 *
 * @example
 * getLengths(["scala", "rust", "ada"]) // => [5, 4, 3]
 */
export const getLengths = (words: readonly string[]): readonly number[] =>
  pipe(
    words,
    RA.map((word) => word.length)
  )

/**
 * 各単語のスコアを取得する。
 *
 * @example
 * getScores(["rust", "java"], score) // => [4, 2]
 */
export const getScores = (
  words: readonly string[],
  wordScore: (word: string) => number
): readonly number[] =>
  pipe(words, RA.map(wordScore))

/**
 * 全ての数値を2倍にする。
 *
 * @example
 * doubleAll([5, 1, 2, 4, 0]) // => [10, 2, 4, 8, 0]
 */
export const doubleAll = (numbers: readonly number[]): readonly number[] =>
  pipe(
    numbers,
    RA.map((n) => n * 2)
  )

/**
 * 全ての数値を負にする。
 *
 * @example
 * negateAll([5, 1, 2, 4, 0]) // => [-5, -1, -2, -4, 0]
 */
export const negateAll = (numbers: readonly number[]): readonly number[] =>
  pipe(
    numbers,
    RA.map((n) => -n)
  )

// =============================================================================
// 4.3 filter - 条件に合う要素を抽出
// =============================================================================

/**
 * 指定した長さ以下の単語をフィルタする。
 *
 * @example
 * filterShortWords(["scala", "rust", "ada"], 4) // => ["rust", "ada"]
 */
export const filterShortWords = (
  words: readonly string[],
  maxLength: number
): readonly string[] =>
  pipe(
    words,
    RA.filter((word) => word.length <= maxLength)
  )

/**
 * 奇数のみをフィルタする。
 *
 * @example
 * filterOdd([5, 1, 2, 4, 0]) // => [5, 1]
 */
export const filterOdd = (numbers: readonly number[]): readonly number[] =>
  pipe(
    numbers,
    RA.filter((n) => n % 2 === 1)
  )

/**
 * 閾値より大きい数値をフィルタする。
 *
 * @example
 * filterLargerThan([5, 1, 2, 4, 0], 4) // => [5]
 */
export const filterLargerThan = (
  numbers: readonly number[],
  threshold: number
): readonly number[] =>
  pipe(
    numbers,
    RA.filter((n) => n > threshold)
  )

/**
 * 指定したスコアより高い単語をフィルタする。
 *
 * @example
 * highScoringWords(["rust", "java"], score, 3) // => ["rust"]
 */
export const highScoringWords = (
  words: readonly string[],
  wordScore: (word: string) => number,
  higherThan: number
): readonly string[] =>
  pipe(
    words,
    RA.filter((word) => wordScore(word) > higherThan)
  )

// =============================================================================
// 4.4 reduce - 累積演算
// =============================================================================

/**
 * 数値の合計を計算する。
 *
 * @example
 * sumAll([5, 1, 2, 4, 100]) // => 112
 */
export const sumAll = (numbers: readonly number[]): number =>
  pipe(
    numbers,
    RA.reduce(0, (acc, n) => acc + n)
  )

/**
 * 全ての文字列の長さの合計を計算する。
 *
 * @example
 * totalLength(["scala", "rust", "ada"]) // => 12
 */
export const totalLength = (words: readonly string[]): number =>
  pipe(
    words,
    RA.reduce(0, (acc, word) => acc + word.length)
  )

/**
 * 全ての文字列中の's'の数の合計を計算する。
 *
 * @example
 * totalSCount(["scala", "haskell", "rust", "ada"]) // => 3
 */
export const totalSCount = (words: readonly string[]): number =>
  pipe(
    words,
    RA.reduce(0, (acc, word) => acc + numberOfS(word))
  )

/**
 * リスト内の最大値を見つける。
 *
 * @example
 * findMax([5, 1, 2, 4, 15]) // => 15
 */
export const findMax = (numbers: readonly number[]): number | undefined => {
  if (numbers.length === 0) return undefined
  return pipe(
    numbers,
    RA.reduce(numbers[0], (max, n) => (n > max ? n : max))
  )
}

/**
 * 全ての単語のスコアの合計を計算する。
 *
 * @example
 * cumulativeScore(["rust", "java"], score) // => 6
 */
export const cumulativeScore = (
  words: readonly string[],
  wordScore: (word: string) => number
): number =>
  pipe(
    words,
    RA.reduce(0, (acc, word) => acc + wordScore(word))
  )

// =============================================================================
// 4.5 関数を返す関数
// =============================================================================

/**
 * n より大きいかを判定する関数を返す。
 *
 * @example
 * largerThan(4)(5) // => true
 * largerThan(4)(3) // => false
 * [5, 1, 2, 4, 0].filter(largerThan(4)) // => [5]
 */
export const largerThan = (n: number): ((i: number) => boolean) =>
  (i) => i > n

/**
 * n で割り切れるかを判定する関数を返す。
 *
 * @example
 * divisibleBy(5)(15) // => true
 * divisibleBy(5)(7) // => false
 * [5, 1, 2, 4, 15].filter(divisibleBy(5)) // => [5, 15]
 */
export const divisibleBy = (n: number): ((i: number) => boolean) =>
  (i) => i % n === 0

/**
 * 長さが n より短いかを判定する関数を返す。
 *
 * @example
 * shorterThan(4)("ada") // => true
 * shorterThan(4)("scala") // => false
 * ["scala", "ada"].filter(shorterThan(4)) // => ["ada"]
 */
export const shorterThan = (n: number): ((s: string) => boolean) =>
  (s) => s.length < n

/**
 * 's' が count より多く含まれるかを判定する関数を返す。
 *
 * @example
 * containsSMoreThan(0)("rust") // => true
 * containsSMoreThan(0)("ada") // => false
 * ["rust", "ada"].filter(containsSMoreThan(0)) // => ["rust"]
 */
export const containsSMoreThan = (count: number): ((s: string) => boolean) =>
  (s) => numberOfS(s) > count

// =============================================================================
// 4.6 Word Scoring の実践的な例
// =============================================================================

/**
 * 'c' を含む単語にボーナスを与える。
 *
 * @example
 * bonus("scala") // => 5
 * bonus("java") // => 0
 */
export const bonus = (word: string): number =>
  word.includes('c') ? 5 : 0

/**
 * 's' を含む単語にペナルティを与える。
 *
 * @example
 * penalty("rust") // => 7
 * penalty("java") // => 0
 */
export const penalty = (word: string): number =>
  word.includes('s') ? 7 : 0

/**
 * スコア + ボーナス を計算する。
 *
 * @example
 * wordScoreWithBonus("scala") // => 8
 * wordScoreWithBonus("java") // => 2
 */
export const wordScoreWithBonus = (word: string): number =>
  score(word) + bonus(word)

/**
 * スコア + ボーナス - ペナルティ を計算する。
 *
 * @example
 * wordScoreWithBonusAndPenalty("java") // => 2
 * wordScoreWithBonusAndPenalty("scala") // => 1
 * wordScoreWithBonusAndPenalty("rust") // => -3
 */
export const wordScoreWithBonusAndPenalty = (word: string): number =>
  score(word) + bonus(word) - penalty(word)

/**
 * スコア関数を受け取り、閾値と単語リストを受け取る関数を返す。
 * カリー化された関数の例。
 *
 * @example
 * const words = ["ada", "haskell", "scala", "java", "rust"]
 * const scorer = highScoringWordsWithThreshold(wordScoreWithBonusAndPenalty)
 * scorer(1)(words) // => ["java"]
 * scorer(0)(words) // => ["ada", "scala", "java"]
 */
export const highScoringWordsWithThreshold =
  (wordScoreFn: (word: string) => number) =>
  (higherThan: number) =>
  (words: readonly string[]): readonly string[] =>
    pipe(
      words,
      RA.filter((word) => wordScoreFn(word) > higherThan)
    )

// =============================================================================
// 4.7 汎用的な高階関数
// =============================================================================

/**
 * map 関数の実装。
 *
 * @example
 * myMap(x => x * 2, [1, 2, 3]) // => [2, 4, 6]
 * myMap(s => s.length, ["a", "bb", "ccc"]) // => [1, 2, 3]
 */
export const myMap = <T, U>(
  func: (item: T) => U,
  lst: readonly T[]
): readonly U[] =>
  lst.map(func)

/**
 * filter 関数の実装。
 *
 * @example
 * myFilter(x => x > 0, [-1, 0, 1, 2]) // => [1, 2]
 * myFilter(s => s.length > 2, ["a", "bb", "ccc"]) // => ["ccc"]
 */
export const myFilter = <T>(
  predicate: (item: T) => boolean,
  lst: readonly T[]
): readonly T[] =>
  lst.filter(predicate)

/**
 * reduce 関数の実装。
 *
 * @example
 * myReduce((acc, x) => acc + x, [1, 2, 3], 0) // => 6
 * myReduce((acc, s) => acc + s.length, ["a", "bb"], 0) // => 3
 */
export const myReduce = <T, U>(
  func: (acc: U, item: T) => U,
  lst: readonly T[],
  initial: U
): U =>
  lst.reduce(func, initial)

// =============================================================================
// 4.8 プログラミング言語の例
// =============================================================================

/** プログラミング言語 */
export interface ProgrammingLanguage {
  readonly name: string
  readonly year: number
}

/** ProgrammingLanguage を作成する */
export const createLanguage = (
  name: string,
  year: number
): ProgrammingLanguage => ({ name, year })

/**
 * 言語を年でソートする（昇順）。
 *
 * @example
 * const langs = [createLanguage("Scala", 2004), createLanguage("Java", 1995)]
 * sortByYear(langs) // => [{name: "Java", year: 1995}, {name: "Scala", year: 2004}]
 */
export const sortByYear = (
  languages: readonly ProgrammingLanguage[]
): readonly ProgrammingLanguage[] =>
  pipe(
    languages,
    RA.sortBy([Ord.contramap((l: ProgrammingLanguage) => l.year)(N.Ord)])
  )

/**
 * 言語名を取得する。
 *
 * @example
 * const langs = [createLanguage("Scala", 2004), createLanguage("Java", 1995)]
 * getNames(langs) // => ["Scala", "Java"]
 */
export const getNames = (
  languages: readonly ProgrammingLanguage[]
): readonly string[] =>
  pipe(
    languages,
    RA.map((l) => l.name)
  )

/**
 * 指定した年より後に作られた言語をフィルタする。
 *
 * @example
 * const langs = [createLanguage("Scala", 2004), createLanguage("Java", 1995)]
 * filterYoungLanguages(langs, 2000) // => [{name: "Scala", year: 2004}]
 */
export const filterYoungLanguages = (
  languages: readonly ProgrammingLanguage[],
  afterYear: number
): readonly ProgrammingLanguage[] =>
  pipe(
    languages,
    RA.filter((l) => l.year > afterYear)
  )
